### ISSUE TITLE
`bridge` CNI plugin in addition to `ptp`

### ISSUE THREAD

**Issue by hellt** [kubernetes-sigs/kind#2444]
**What would you like to be added**:
I would like to propose a feature request to add a customization option for kindnet to be able to create not only `veth` pairs in the rootNS of the node which is done with `ptp` CNI plugin, but also to use `bridge` CNI plugin for that.

**Why is this needed**:
Containerized Network OS often comes with userplane dataplane functionality which may not play well with the custom static routes kindnet installs in PODs namespace.

For example Nokia SR Linux container implements its own dataplane in user space, and it uses custom namespaces to separate mgmt plane from the dataplane.
Long story short, the static routes for podCIDR that kindnetd installs in pod's root namespace have no effect on the said container, since its management plane is in the other namespace.
This led to a non-working inter-pod communicatoin between said pods that I showed to @aojea the other day here https://twitter.com/Itsuugo/status/1432820443321409546

By using `bridge` plugin though, it seems the problem will get solved, as the L2 domain will be common for all the pods inside the node and the ARP will work.

**Comment by aojea** [comment]
I don't think that we'll add it as part of kind, but I'll try to figure out a simple solution
/assign

**Comment by hellt** [comment]
just a quick update
by installing a bridge cni plugin on a kind node and replacing `ptp` with `bridge` in `/etc/cni/net.d/10-kindnet.conflist` I managed to get inter-pod connectivity when pods are scheduled on the same node

**Comment by aojea** [comment]
> just a quick update
> by installing a bridge cni plugin on a kind node and replacing `ptp` with `bridge` in `/etc/cni/net.d/10-kindnet.conflist` I managed to get inter-pod connectivity when pods are scheduled on the same node

Yeah, those sr pods consider the Mgmt interface attached to a L2 network, hence they arp to reach pods on the same node.
Ptp is pure routing, and the SR should use the default gateway that is the other end of the veth

**Comment by aojea** [comment]
yeah, well, I was trying to point out this specific problem, that srlinux doesn't use the default gw and just arps, so it fails to reach pods on the same subnet with ptp plugins. The technical details of the ptp hack and your explanation is absolutely correct
https://github.com/containernetworking/plugins/blob/189d0c06aa6da0c4f052d3831b091e6ea79e6675/plugins/main/ptp/ptp.go#L54-L63

My point is that these srlinux pods will require the cluster to always use a CNI with bridge , and this is a strong requirement for a cluster, I can't say what weave, calico and other plugins use by default

**Comment by BenTheElder** [comment]
FWIW you have the option to disable the built in CNI and install your own implementation we just don't support them other than leaving the option readily available.

I don't think we want to make kindnetd more complex / parameterized or expose the implementation details and I don't think you can actually expect to have a bridge in Kubernetes as a pod. For example IIRC most GKE nodes are also ptp.

**Comment by LimeHat** [comment]
> FWIW you have the option to disable the built in CNI and install your own implementation we just don't support them other than leaving the option readily available.

Absolutely (I have no opinion on the initial request), I'll leave it up to @hellt
> I don't think you can actually expect to have a bridge in Kubernetes as a pod. For example IIRC most GKE nodes are also ptp.

Probably it's not worth going into a long discussion, but I'd say that
1. we don't expect a bridge per se; more like we rely on the fact that CNI will not use any tricky routing setups
2. we understand that it may not happen in some corner cases as with PTP, and, in my opinion, that's fine (if not, we'll have to work something out on our end, of course)
3. AFAIK GKE does bridging flawlessly and its part of their [architecture](https://cloud.google.com/kubernetes-engine/docs/concepts/network-overview)<sup>1</sup> ![gke](https://cloud.google.com/kubernetes-engine/images/networking-overview_single-node.png)
<sup>1</sup>there will be slight differences between dataplane v1 and v2, but not the point really

**Comment by hellt** [comment]
As an update to this, the following was required to make default kind cluster to work with PODs which are not expecting routes to be installed in their def netns

https://gist.github.com/hellt/806e6cc8d6ae49e2958f11b4a1fc3091

(not all of this applies to pure kind, as some of the steps were touching the meshnet-cni)

**Comment by aojea** [comment]
Thanks for following up, glad you've found a solution ... I didn't have time for this sorry

**Comment by hellt** [comment]
@aojea oh, super!
How do I get to know which kind release will use kindnet 0.8+ so that I can test this?

**Comment by aojea** [comment]
> @aojea oh, super! How do I get to know which kind release will use kindnet 0.8+ so that I can test this?

We've been discussing this yesterday in slack https://kubernetes.slack.com/archives/CEKK1KTN2/p1634767605303200

When we started to discuss the CNI approach in KIND we made a decision to avoid leaking CNI internal details into KIND, so we choose a minimalistic CNI plugin (kindnet) and option for people to install their own CNI with `disableDefaultCNI`

Since there are other projects that to use kindnet, I'm maintaining a copy https://github.com/aojea/kindnet/ outside of kind with some modifications to allow to run outside of KIND.
The workflow will be to install the cluster without the embedded kindnet:
```
kind: Cluster
apiVersion: kind.x-k8s.io/v1alpha4
networking:
  disableDefaultCNI: true
```
and after installation, install the external kindnet.

**Comment by hellt** [comment]
@aojea  got it
Do you know if it is also possible to disable the offload on that bridge as this, again, can mess up containerized routing systems which do not expect to receive packets with incorrect checksums.
and checksums by default are fake on the same host

with ethtool: https://gist.github.com/hellt/806e6cc8d6ae49e2958f11b4a1fc3091#file-bridge-sh-L28
in Go: https://github.com/srl-labs/containerlab/blob/96a107f2b3fd1c050ecaf7c3f237266d7ea3a4f1/utils/ethtool.go#L41

/cc @marcushines

**Comment by jseguillon** [comment]
Hi there,
As a user of KubeVirt inside Kind, I would be super interrested in having a solution for local bridge as second network for my VM Pods. KubeVirt does not support ptp and I come to same dataplane need as expressed by issue author.

Could we work on some "local-kind-bridge" CNI with any needed customization for a working bridge within on one and only one node  and make it available by default ?

**Comment by BenTheElder** [comment]
https://github.com/kubernetes-sigs/kind/issues/2444#issuecomment-948528390

**Comment by hellt** [comment]
I'd like to provide a workaround created by @aojea for completeness


## Installing a cluster with a custom kindnet CNI

Install a kind cluster with a config file that disables default ptp-based kindnet CNI from being applied
```bash
kind create cluster --config=kind-no-def-cni.yml --name=cluster1
```

`kind-no-def-cni.yml` contains kind config options to exclude default CNI from being deployed:

```yml
kind: Cluster
apiVersion: kind.x-k8s.io/v1alpha4
networking:
  # the default CNI will not be installed
  disableDefaultCNI: true
```

Once deployed, apply bridge-based CNI plugin from this manifest

```bash
kubectl apply -f https://raw.githubusercontent.com/aojea/kindnet/master/install-kindnet-bridge.yaml
```

<details><summary>contents of <code>install-kindnet-bridge.yaml</code></summary>
<p>

```yaml
---
kind: ClusterRole
apiVersion: rbac.authorization.k8s.io/v1
metadata:
  name: kindnet
rules:
  - apiGroups:
      - ""
    resources:
      - nodes
    verbs:
      - list
      - watch
      - patch
  - apiGroups:
     - ""
    resources:
      - configmaps
    verbs:
      - get
---
kind: ClusterRoleBinding
apiVersion: rbac.authorization.k8s.io/v1
metadata:
  name: kindnet
roleRef:
  apiGroup: rbac.authorization.k8s.io
  kind: ClusterRole
  name: kindnet
subjects:
- kind: ServiceAccount
  name: kindnet
  namespace: kube-system
---
apiVersion: v1
kind: ServiceAccount
metadata:
  name: kindnet
  namespace: kube-system
---
apiVersion: apps/v1
kind: DaemonSet
metadata:
  name: kindnet
  namespace: kube-system
  labels:
    tier: node
    app: kindnet
    k8s-app: kindnet
spec:
  selector:
    matchLabels:
      app: kindnet
  template:
    metadata:
      labels:
        tier: node
        app: kindnet
        k8s-app: kindnet
    spec:
      hostNetwork: true
      tolerations:
      - operator: Exists
        effect: NoSchedule
      serviceAccountName: kindnet
      initContainers:
      - name: install-cni-bin
        image: aojea/kindnetd
        command: ['sh', '-c', 'cp -f /opt/cni/bin/* /cni/ ']
        volumeMounts:
        - name: cni-bin
          mountPath: /cni
      containers:
      - name: kindnet-cni
        image: aojea/kindnetd
        env:
        - name: HOST_IP
          valueFrom:
            fieldRef:
              fieldPath: status.hostIP
        - name: POD_IP
          valueFrom:
            fieldRef:
              fieldPath: status.podIP
        - name: CNI_BRIDGE
          value: "true"
        volumeMounts:
        - name: cni-cfg
          mountPath: /etc/cni/net.d
        - name: xtables-lock
          mountPath: /run/xtables.lock
          readOnly: false
        - name: lib-modules
          mountPath: /lib/modules
          readOnly: true
        resources:
          requests:
            cpu: "100m"
            memory: "50Mi"
          limits:
            cpu: "100m"
            memory: "50Mi"
        securityContext:
          privileged: false
          capabilities:
            add: ["NET_RAW", "NET_ADMIN"]
      volumes:
      - name: cni-bin
        hostPath:
          path: /opt/cni/bin
          type: DirectoryOrCreate
      - name: cni-cfg
        hostPath:
          path: /etc/cni/net.d
          type: DirectoryOrCreate
      - name: xtables-lock
        hostPath:
          path: /run/xtables.lock
          type: FileOrCreate
      - name: lib-modules
        hostPath:
          path: /lib/modules
---
```
</p>
</details>

Ensure, that cluster pods are up and running.

**Comment by hellt** [comment]
as was brought in kind slack channel, this is not working as of today. kube-system core-dns pods fail to communicate with api server - https://kubernetes.slack.com/archives/CEKK1KTN2/p1656000803486879?thread_ts=1634767605.303200&cid=CEKK1KTN2

**Comment by BenTheElder** [comment]
I think these advanced use cases are covered by the aojea/kindnetd repo + disableDefaultCNI.

The default CNI is a happy path for simple network users. For power users disabling the built in CNI offering and installing alternative / modified versions is the supported path.
